nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py

```diff
@@ -1284,7 +1394,31 @@ def dummy():
             micro_batch_size=global_batch_per_gpu // num_micro_batches_before_decode,
             data_parallel_size=parallel_state.get_data_parallel_world_size(),
         )
-        return predicted_tokens_dec, predicted_log_probs
+
+        if beam_search and beam_size > 1:
+            if keep_only_best_tokens:
+                len_penalties = compute_beam_search_len_penalty(decoder_seq_lengths, 0)
+                scores = scores / len_penalties
+                scores = scores.view(-1, beam_size)
+                best_ids = torch.argmax(scores, dim=1, keepdim=True)
+                scores = scores * len_penalties.view(-1, beam_size)
+                scores = scores.gather(1, best_ids)
+                best_tokens = best_ids.repeat(1, predicted_tokens_dec.size(1)).unsqueeze(1)
+                predicted_tokens_dec = (
+                    predicted_tokens_dec.view(batch_size, beam_size, -1).gather(1, best_tokens).squeeze(1)
+                )
+                predicted_log_probs = (
+                    predicted_log_probs.view(batch_size, beam_size, -1).gather(1, best_tokens[:, :, 1:]).squeeze(1)
+                )
+            else:
+                predicted_tokens_dec = predicted_tokens_dec.view(batch_size, beam_size, -1)
+                predicted_log_probs = predicted_log_probs.view(batch_size, beam_size, -1)
+                scores = scores.view(-1, beam_size)
+
+        if beam_search and return_scores:
+            return predicted_tokens_dec, predicted_log_probs, scores
+        else:
+            return predicted_tokens_dec, predicted_log_probs
```

Comment:
Not priority: maybe we can merge conditions and run beam_search with beam_size=1 as greedy in the future?

Reply:
I would suggest to keep them separated since greedy is faster. Does it make sense?